src/ucp/tag/tag_send.c

```diff
@@ -185,100 +200,125 @@ ucp_tag_send_inline(ucp_ep_h ep, const void *buffer, size_t count,
     return status;
 }
 
-
 UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_tag_send_nb,
                  (ep, buffer, count, datatype, tag, cb),
                  ucp_ep_h ep, const void *buffer, size_t count,
                  uintptr_t datatype, ucp_tag_t tag, ucp_send_callback_t cb)
 {
-    ucs_status_t status;
-    ucp_request_t *req;
-    ucs_status_ptr_t ret;
+    ucp_request_param_t param = {
+        .op_attr_mask = UCP_OP_ATTR_FIELD_DATATYPE | UCP_OP_ATTR_FIELD_CALLBACK,
+        .cb.send      = (ucp_send_nbx_callback_t)cb,
+        .datatype     = datatype
+    };
 
-    UCP_CONTEXT_CHECK_FEATURE_FLAGS(ep->worker->context, UCP_FEATURE_TAG,
-                                    return UCS_STATUS_PTR(UCS_ERR_INVALID_PARAM));
-    UCP_WORKER_THREAD_CS_ENTER_CONDITIONAL(ep->worker);
-
-    ucs_trace_req("send_nb buffer %p count %zu tag %"PRIx64" to %s cb %p",
-                  buffer, count, tag, ucp_ep_peer_name(ep), cb);
+    return ucp_tag_send_nbx(ep, buffer, count, tag, &param);
+}
 
-    status = UCS_PROFILE_CALL(ucp_tag_send_inline, ep, buffer, count,
-                              datatype, tag);
-    if (ucs_likely(status != UCS_ERR_NO_RESOURCE)) {
-        ret = UCS_STATUS_PTR(status); /* UCS_OK also goes here */
-        goto out;
+UCS_PROFILE_FUNC(ucs_status_t, ucp_tag_send_nbr,
+                 (ep, buffer, count, datatype, tag, request),
+                 ucp_ep_h ep, const void *buffer, size_t count,
+                 uintptr_t datatype, ucp_tag_t tag, void *request)
+{
+    ucp_request_param_t param = {
+        .op_attr_mask = UCP_OP_ATTR_FIELD_DATATYPE | UCP_OP_ATTR_FIELD_REQUEST |
+                        UCP_OP_ATTR_FLAG_NO_ZCOPY,
+        .datatype     = datatype,
+        .request      = request
+    };
+    ucs_status_ptr_t status;
+
+    status = ucp_tag_send_nbx(ep, buffer, count, tag, &param);
+
+    if (ucs_likely(status == UCS_OK)) {
+        return UCS_OK;
     }
 
-    req = ucp_request_get(ep->worker);
-    if (req == NULL) {
-        ret = UCS_STATUS_PTR(UCS_ERR_NO_MEMORY);
-        goto out;
+    if (ucs_unlikely(UCS_PTR_IS_ERR(status))) {
```

Comment:
can use `UCS_PTR_STATUS` here and in other places

Reply:
right, but here we check for error